packages/teleterm/src/ui/services/clusters/clustersService.ts

```diff
@@ -13,6 +13,7 @@ import {
   Cluster,
```

Comment:
First of all, love pair programming so sign me up. 
I think making autocomplete async is fine although I'd rather not leave it broken. If anything, I'd rather just leave it inefficient by updating the map with new entries. I may be naive here, but it seems as simple as making a new function similar to `updateMap`
```js
const helpers = {
  updateMap<T extends { uri: string }>(
    parentUri = '',
    map: Map<string, T>,
    received: T[]
  ) {
    // delete all entries under given uri
    for (let k of map.keys()) {
      if (routing.ensureClusterUri(k) === parentUri) {
        map.delete(k);
      }
    }

    received.forEach(s => map.set(s.uri, s));
  },
};
```

but just skipping the delete phase. This would make it work and be very little code to remove in the autocomplete async PR. Either way, I don't mind.

Good news I think is, if we implement everything using the same async functions (`fetchServers`, `fetchKubes`, etc), once livecache comes into play, we _should_ be able to just update those async functions to pull from the cache instead and the rest of the app would work.


> remove resource counts in DocumentCluster (they're going to be present at the bottom of the table anyway)

Yes. Agreed

Reply:
The problems with just updating the cache when server-side requests come in are:

* The autocomplete might miss some entries if there was no server-side search which would return those entries.
   * That's what I was referring to when I said we'd leave the autocomplete slightly broken, I should've been more clear here.
* The autocomplete is still going to be unusable on clusters with a lot of resources as it does all filtering within the main browser thread.

> I think making autocomplete async is fine although I'd rather not leave it broken. If anything, I'd rather just leave it inefficient by updating the map with new entries.

I don't understand the distinction between broken and inefficient that you used here, what you wrote seems to imply that making the autocomplete async == leaving it broken? ;)

When I said that we'd leave it slightly broken, I referred to a situation where we would leave it sync, depending on `ClustersService` cache.

> Good news I think is, if we implement everything using the same async functions (`fetchServers`, `fetchKubes`, etc), once livecache comes into play, we _should_ be able to just update those async functions to pull from the cache instead and the rest of the app would work.

Right. And I imagine, though I'm not 100% sure, that whatever solution we're going to come up with for the live cache might look significantly different than what we have now so that's another argument for ripping out resource management from `ClustersService` for now.

Reply:
I misunderstood broken. I didn't apply "doesn't work with large clusters" to the broken adjective and only thought of it missing entries. What you spoke of originally makes more sense to me now and I'm onboard 👍

Reply:
New working hours are fine, I'll make it work no problem at all. I think that although priorities have changed for the quarter, since this (to me) is pretty close to done (advanced search), might as well get it out. If we think that changing autocomplete is required for this to go out (which to be clear, I do), I think the that work is worth it. A couple hours in the grand scheme of the quarter shouldn't affect the goals. I'll setup some time next week on our calendar.